IpmiFeaturePkg/Library/IpmiSelLib/IpmiSelLib.c

```diff
@@ -0,0 +1,499 @@
+/** @file
+  Implements the SEL library. This library simplifies using the IPMI sel
+  functionality.
+
+  Copyright (c) Microsoft Corporation
+  PDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Uefi.h>
+#include <Library/DebugLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/TimerLib.h>
+
+#include <IndustryStandard/Ipmi.h>
+#include <Library/IpmiBaseLib.h>
+#include <Library/IpmiSelLib.h>
+
+//
+// All SEL entries are the same size and exactly 16 bytes.
+//
+
+#define SEL_ENTRY_SIZE  (16)
+STATIC_ASSERT (
+  sizeof (IPMI_SEL_EVENT_RECORD_DATA) == SEL_ENTRY_SIZE,
+  "Unexpected SEL entry size!"
+  );
+
+STATIC_ASSERT (
+  sizeof (IPMI_TIMESTAMPED_OEM_SEL_RECORD_DATA) == SEL_ENTRY_SIZE,
+  "Unexpected SEL entry size!"
+  );
+
+STATIC_ASSERT (
+  sizeof (IPMI_NON_TIMESTAMPED_OEM_SEL_RECORD_DATA) == SEL_ENTRY_SIZE,
+  "Unexpected SEL entry size!"
+  );
+
+/**
+  Converts a IPMI completion code to a EFI status code.
+
+  @param[in]  CompCode    The IPMI completion code to convert.
+
+  @retval     EFI_STATUS best representing the completion code.
+**/
+EFI_STATUS
+EFIAPI
+IpmiCompCodeToEfiStatus (
+  IN UINT8  CompCode
+  )
+{
+  switch (CompCode) {
+    case IPMI_COMP_CODE_NORMAL:
+      return EFI_SUCCESS;
+
+    case IPMI_COMP_CODE_INVALID_COMMAND:
+      return EFI_UNSUPPORTED;
+
+    case IPMI_COMP_CODE_INVALID_FOR_GIVEN_LUN:
+    case IPMI_COMP_CODE_INVALID_DATA_FIELD:
+    case IPMI_COMP_CODE_COMMAND_ILLEGAL:
+      return EFI_INVALID_PARAMETER;
+
+    case IPMI_COMP_CODE_INVALID_REQUEST_DATA_LENGTH:
+      return EFI_BAD_BUFFER_SIZE;
+
+    case IPMI_COMP_CODE_NOT_PRESENT:
+    case IPMI_COMP_CODE_OUT_OF_RANGE:
+      return EFI_NOT_FOUND;
+
+    case IPMI_COMP_CODE_TIMEOUT:
+      return EFI_TIMEOUT;
+
+    case IPMI_COMP_CODE_OUT_OF_SPACE:
+      return EFI_OUT_OF_RESOURCES;
+
+    case IPMI_COMP_CODE_BMC_INIT_IN_PROGRESS:
+      return EFI_NOT_READY;
+
+    case IPMI_COMP_CODE_INSUFFICIENT_PRIVILEGE:
+      return EFI_SECURITY_VIOLATION;
+
+    default:
+      return EFI_PROTOCOL_ERROR;
+  }
+}
+
+/**
+  Adds a generic entry to the SEL.
+
+  @param[in]      Entry       The entry to be added to the SEL.
+  @param[in]      EntrySize   The size of the entry in bytes.
+  @param[in,out]  RecordId    If provided, receives the record ID of the entry.
+
+  @retval   EFI_SUCCESS         The entry was successfully added.
+  @retval   EFI_PROTOCOL_ERROR  Unexpected result size.
+  @retval   Other               The IPMI base library returned an error.
+**/
+EFI_STATUS
+EFIAPI
+IpmiAddSelEntry (
+  IN VOID        *Entry,
+  IN UINT32      EntrySize,
+  IN OUT UINT16  *RecordId OPTIONAL
+  )
+{
+  EFI_STATUS                   Status;
+  UINT32                       DataSize;
+  IPMI_ADD_SEL_ENTRY_RESPONSE  Response;
+
+  ASSERT (EntrySize == SEL_ENTRY_SIZE);

```

Comment:
Should we have a non-assert check for production FW? Or, is the expectation that if the entry size is too long, we'll just send `SEL_ENTRY_SIZE` bytes?

Reply:
i'll do one better. im doing away with the VOID * in favor of a unified strucure.